apollo-runtime/src/main/java/com/apollographql/apollo/internal/batch/BatchPoller.kt

```diff
@@ -0,0 +1,82 @@
+package com.apollographql.apollo.internal.batch
+
+import com.apollographql.apollo.api.ScalarTypeAdapters
+import com.apollographql.apollo.api.internal.ApolloLogger
+import com.apollographql.apollo.exception.ApolloException
+import com.apollographql.apollo.interceptor.ApolloInterceptor
+import okhttp3.Call
+import okhttp3.HttpUrl
+import java.util.*
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import kotlin.collections.ArrayList
+
+/**
+ * Class responsible for holding the queue of requests, and execute them as a batch every X ms,
+ * as defined by the BatchConfig.
+ */
+class BatchPoller(
+    private val serverUrl: HttpUrl,
+    private val httpCallFactory: Call.Factory,
+    private val scalarTypeAdapters: ScalarTypeAdapters,
+    private val logger: ApolloLogger,
+    private val batchConfig: BatchConfig,
+    private val dispatcher: Executor,
+    private val scheduledExecutorService: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
+) {
+
+  private val queryQueue: ConcurrentLinkedQueue<QueryToBatch> = ConcurrentLinkedQueue()
+  private var pollDisposable: ScheduledFuture<*>? = null
+
+  fun start() {
+    stop()
+    pollDisposable = scheduledExecutorService.scheduleAtFixedRate({
+      maybeExecuteBatchQuery(queryQueue)
+    }, 0, batchConfig.batchIntervalMs, TimeUnit.MILLISECONDS)
+  }
+
+  fun stop() {
+    pollDisposable?.cancel(true)
+    pollDisposable = null
+  }
+
+  fun enqueue(query: QueryToBatch) {
+    if (pollDisposable == null) {
+      throw ApolloException("Trying to batch queries without calling ApolloClient.startBatchPoller() first")
+    }
+    queryQueue.add(query)
+    logger.d("Enqueued Query: ${query.request.operation.name().name()} for batching")
+  }
+
+  fun removeFromQueue(query: QueryToBatch) {
+    queryQueue.remove(query)
+  }
+
+  private fun maybeExecuteBatchQuery(queryQueue: Queue<QueryToBatch>) {
+    if (queryQueue.isEmpty()) {
+      return
+    }
+    // copy and clear the current queue
+    val queryList = ArrayList(queryQueue)
+    queryQueue.clear()
+    // split into batches
+    val batches = queryList.chunked(batchConfig.maxBatchSize)
+    logger.d("Executing ${queryList.size} Queries in ${batches.size} Batch(es)")
+    for (batch in batches) {
+      dispatcher.execute { createBatchHttpCall(batch).execute() }
```

Comment:
Might be worth eagerly sending the `maxBatchSize` operations as soon as they're enqueued to save some latency instead of doing it from the timer?